libs/ssd1327/ssd1327.cpp

```diff
@@ -0,0 +1,172 @@
+
+    command(0xbc);    // set pre-charge voltage
+    command(0x08);    // 0x08 = VCCOM
+
+    command(0xd5);    // Function Selection B
+    command(0x62);    // 0x62 = 0b01100010
+
+    command(0xfd);    // Select command lock
+    command(0x12);    // 0x12 = 0b00010010 (unlocked)
+
+    command(0xAF);    //Turn on the OLED display
+
+    // erase lower half (unused) part of display
+    clear();          // set buffer (half display size) to all zero
+    command(0x75);    //set row address for lower half
+    command(0x40);    //start row   64
+    command(0x7f);    //end row   127
+    flush();          // write buffer to selected row range
+    command(0x75);    //set row address back to upper half
+    command(0x0);    //start row   64
+    command(0x3f);    //end row   127
+}
+
+void Ssd1327::clear()
+{
+    for (size_t i = 0; i < width * height / 8; i++) {
+        m_buffer[i] = 0;
+    }
+}
+
+void Ssd1327::flush()
+{
+    uint8_t *buffer = m_buffer;
+
+    for (uint8_t page = 0; page < 8; page++) {
+        // write data
+        digitalWrite(m_dc, HIGH);
+
+        for (size_t i = 0; i < width; i++) {
+            // change to 4bit grayscale
+            // each bit should be replaced by b0000 or b1111 depending on its value
+            for (uint8_t j = 0; j < 4; j++) {    // iterate over each 2bit pair in single byte of buffer
+                uint8_t temp_buffer = buffer[i] >> (j*2);
+                // 1. mask to lower two bits
+                temp_buffer &= 0b00000011;
+                // 2. replace 2bits by grayscale byte
+                switch(temp_buffer) {
+                    case 0b00:
+                        SPI.transfer(0b00000000);
+                        break;
+                    case 0b01:
+                        SPI.transfer(0b00001111);
+                        break;
+                    case 0b10:
+                        SPI.transfer(0b11110000);
+                        break;
+                    case 0b11:
+                        SPI.transfer(0b11111111);
+                        break;
+                };
+            }
+        }
+        buffer += width;
+    }
+}
+
+void Ssd1327::draw_pixel(uint8_t x, uint8_t y)
+{
+    if (x > width || y > height)
+        return;
+
+    m_buffer[(x / 8) + y * (width / 8)] |= 1 << (x % 8);    // example: first byte 0xFF equals horizontal line starting upper left and 8px length --> 9px length woud imply added second byte = b1 (LSBF)
+}
+
+void Ssd1327::draw_bitmap(uint8_t x, uint8_t y, Bitmap&& bitmap)
+{
+    uint8_t byte_width = (bitmap.width + 7) / 8;
+
+    for (uint8_t j = 0; j < bitmap.height; j++){
+        for (uint8_t i = 0; i < bitmap.width; i ++){
+            // Seems stupid to read the same byte over and over again ...
+            if (pgm_read_byte(bitmap.data + j * byte_width + i / 8) & (128 >> (i & 7))) {
+                draw_pixel(x+i, y+j);
+            }
+        }
+    }
+}
```

Comment:
I guess, we can make this function non-virtual and implement directly in the main interface class. It's the same regardless of the controller because it just depends on `draw_pixel`-